signer/src/transaction_signer.rs

```diff
@@ -127,71 +131,91 @@ pub struct TxSignerEventLoop<Network, Storage, BlocklistChecker, Rng> {
     pub network_kind: bitcoin::Network,
     /// Random number generator used for encryption
     pub rng: Rng,
-    #[cfg(feature = "testing")]
-    /// Optional channel to communicate progress usable for testing
-    pub test_observer_tx: Option<tokio::sync::mpsc::Sender<TxSignerEvent>>,
 }
 
-/// Event useful for tests
-#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
-pub enum TxSignerEvent {
-    /// Received a deposit decision
-    ReceivedDepositDecision,
-    /// Received a withdrawal decision
-    ReceivedWithdrawalDecision,
-}
-
-impl<N, S, B, Rng> TxSignerEventLoop<N, S, B, Rng>
+impl<C, N, B, Rng> TxSignerEventLoop<C, N, B, Rng>
 where
+    C: Context,
     N: network::MessageTransfer,
     B: blocklist_client::BlocklistChecker,
-    S: storage::DbRead + storage::DbWrite + Send + Sync,
     Rng: rand::RngCore + rand::CryptoRng,
 {
     /// Run the signer event loop
     #[tracing::instrument(skip(self))]
     pub async fn run(mut self) -> Result<(), Error> {
-        loop {
-            tokio::select! {
-                result = self.block_observer_notifications.changed() => {
-                    match result {
-                        Ok(()) => self.handle_new_requests().await?,
-                        Err(_) => {
-                            tracing::info!("block observer notification channel closed");
-                            break;
-                        }
+        let mut signal_rx = self.context.get_signal_receiver();
+        let mut term = self.context.get_termination_handle();
+
+        // TODO: We should really split these operations out into two separate
+        // main run-loops since they don't have anything to do with eachother.
+        //
+        // We run the event loop like this because `tokio::select!()` could
+        // potentially kill either `handle_new_requests()` or `handle_signer_message()`
+        // in the middle of processing if they end-up running concurrently and
+        // the other one finishes first.
+        let run_task = async {
+            loop {
+                // First we empty the signal channel subscription, checking for
+                // new Bitcoin block observed events. It doesn't matter how many
+                // of these we get, we only care if it has happened. It's also
+                // important that we empty this channel as quickly as possible
+                // to avoid un-processed messagages being dropped.
+                let mut new_block_observed = false;
+                while let Ok(signal) = signal_rx.try_recv() {
+                    if let SignerSignal::Event(SignerEvent::BitcoinBlockObserved) = signal {
+                        new_block_observed = true;
                     }
```

Comment:
Ahh right, it doesn't care about any other internal messages besides the termination signal message. And everything else comes over the network.

Reply:
Well this one only cares about `BitcoinBlockObserved`, termination channel is a separate `watch` channel, and then yeah just below is the stuff coming over the network. So these two operations run in the same loop, and the `select!` below handles killing it when it dies or termination is signalled.

I did try putting these into two separate futures with their own loops but `&mut` quickly became a problem and I went this route instead to not spend time on that.

Reply:
No it makes sense. After reading your comments I couldn't think of a way to do it as separate functions.